File: components/gmaps-flow/version-1.0/google-map-flow/src/main/resources/META-INF/resources/frontend/google-map-rectangle.js
/** 
@license
Copyright 2010-2020 Neotropic SAS <contact@neotropic.co>.

Licensed under the Apache License, Version 2.0 (the "License"); 
you may not use this file except in compliance with the License. 
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software 
distributed under the License is distributed on an "AS IS" BASIS, 
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
See the License for the specific language governing permissions and 
limitations under the License.
*/

import {html, PolymerElement} from '@polymer/polymer/polymer-element.js';

/**
 * `<google-map-rectangle>` Rectangle overlay.
 * @customElement
 * @polymer
 * 
 */
class GoogleMapRectangle extends PolymerElement {
  static get is() {
    return 'google-map-rectangle';
  }

  static get template() {
    return html`
      <style>
        :host {
          display: block;
        }
      </style>
    `;
  }

  static get properties() {
    return {
      /**
       * Rectangle bounds
       * @type {{east: number, north: number, south: number, west: number}}
       */
      bounds: {
        type: Object,
        observer: '_boundsChanged'
      },
      clickable: {
        type: Boolean,
        value: true,
        observer: '_clickableChanged'
      },
      draggable: {
        type: Boolean,
        value: false,
        observer: '_draggableChanged'
      },
      editable: {
        type: Boolean,
        value: false,
        observer: '_editableChanged'
      },
      fillColor: {
        type: String,
        value: '#FF0000',
        observer: '_fillColorChanged'
      },
      fillOpacity: {
        type: Number,
        value: 0.35,
        observer: '_fillOpacityChanged'
      },
      strokeColor: {
        type: String,
        value: '#FF0000',
        observer: '_strokeColorChanged'
      },
      strokeOpacity: {
        type: Number,
        value: 0.8,
        observer: '_strokeOpacityChanged'
      },
      strokePosition: {
        type: String,
        observer: '_strokePositionChanged'
      },
      strokeWeight: {
        type: Number,
        value: 2,
        observer: '_strokeWeightChanged'
      },
      visible: {
        type: Boolean,
        value: true,
        observer: '_visibleChanged'
      },
      zIndex: {
        type: Number,
        observer: '_zIndexChanged'
      }
    };
  }
  /**
   * @return {google.maps.MVCObject} The google.maps.Rectangle
   */
  getMVCObject() {
    return this.rectangle;
  }
  /**
   * @param {google.maps.Map} map 
   */
  added(map) {
    this.rectangle = new google.maps.Rectangle({
      bounds: this.bounds,
      clickable: this.clickable,
      draggable: this.draggable,
      editable: this.editable,
      fillColor: this.fillColor,
      fillOpacity: this.fillOpacity,
      strokeColor: this.strokeColor,
      strokeOpacity: this.strokeOpacity,
      strokeWeight: this.strokeWeight,
      map: map,
      visible: this.visible
    });
    if (this.strokePosition)
      this._strokePositionChanged(this.strokePosition);
    if (this.zIndex)
      this._zIndexChanged(this.zIndex);
    
    this.rectangle.addListener('bounds_changed', event => {
      this.bounds = this.rectangle.getBounds().toJSON();
      this.dispatchEvent(new CustomEvent('rectangle-bounds-changed'));
    });
    this.rectangle.addListener('click', event => {
      this.dispatchEvent(new CustomEvent('rectangle-click', 
        {detail: {lat: event.latLng.lat(), lng: event.latLng.lng()}}
      ));
    });
    this.rectangle.addListener('dblclick', event => {
      this.dispatchEvent(new CustomEvent('rectangle-dbl-click', 
        {detail: {lat: event.latLng.lat(), lng: event.latLng.lng()}}
      ));
    });
    this.rectangle.addListener('drag', event => {
      this.dispatchEvent(new CustomEvent('rectangle-drag', 
        {detail: {lat: event.latLng.lat(), lng: event.latLng.lng()}}
      ));
    });
    this.rectangle.addListener('dragend', event => {
      this.dispatchEvent(new CustomEvent('rectangle-drag-end', 
        {detail: {lat: event.latLng.lat(), lng: event.latLng.lng()}}
      ));
    });
    this.rectangle.addListener('dragstart', event => {
      this.dispatchEvent(new CustomEvent('rectangle-drag-start', 
        {detail: {lat: event.latLng.lat(), lng: event.latLng.lng()}}
      ));
    });
    this.rectangle.addListener('mousedown', event => {
      this.dispatchEvent(new CustomEvent('rectangle-mouse-down', 
        {detail: {lat: event.latLng.lat(), lng: event.latLng.lng()}}
      ));
    });
    this.rectangle.addListener('mousemove', event => {
      this.dispatchEvent(new CustomEvent('rectangle-mouse-move', 
        {detail: {lat: event.latLng.lat(), lng: event.latLng.lng()}}
      ));
    });
    this.rectangle.addListener('mouseout', event => {
      this.dispatchEvent(new CustomEvent('rectangle-mouse-out', 
        {detail: {lat: event.latLng.lat(), lng: event.latLng.lng()}}
      ));
    });
    this.rectangle.addListener('mouseover', event => {
      this.dispatchEvent(new CustomEvent('rectangle-mouse-over', 
        {detail: {lat: event.latLng.lat(), lng: event.latLng.lng()}}
      ));
    });
    this.rectangle.addListener('mouseup', event => {
      this.dispatchEvent(new CustomEvent('rectangle-mouse-up', 
        {detail: {lat: event.latLng.lat(), lng: event.latLng.lng()}}
      ));
    });
    this.rectangle.addListener('rightclick', event => {
      this.dispatchEvent(new CustomEvent('rectangle-right-click', 
        {detail: {lat: event.latLng.lat(), lng: event.latLng.lng()}}
      ));
    });
  }

  removed() {
    if (this.rectangle)
      this.rectangle.setMap(null);
  }

  _boundsChanged(newValue, oldValue) {
    if (this.rectangle)
      this.rectangle.setOptions({bounds: newValue});
  }

  _clickableChanged(newValue, oldValue) {
    if (this.rectangle)
      this.rectangle.setOptions({clickable: newValue});
  }

  _draggableChanged(newValue, oldValue) {
    if (this.rectangle)
      this.rectangle.setOptions({draggable: newValue});
  }

  _editableChanged(newValue, oldValue) {
    if (this.rectangle)
      this.rectangle.setOptions({editable: newValue});
  }

  _fillColorChanged(newValue, oldValue) {
    if (this.rectangle)
      this.rectangle.setOptions({fillColor: newValue});
  }

  _fillOpacityChanged(newValue, oldValue) {
    if (this.rectangle)
      this.rectangle.setOptions({fillOpacity: newValue});
  }

  _strokeColorChanged(newValue, oldValue) {
    if (this.rectangle)
      this.rectangle.setOptions({strokeColor: newValue});
  }

  _strokeOpacityChanged(newValue, oldValue) {
    if (this.rectangle)
      this.rectangle.setOptions({strokeOpacity: newValue});
  }

  _strokePositionChanged(newValue, oldValue) {
    if (this.rectangle) {
      switch(newValue) {
        case 'center':
          this.rectangle.setOptions({strokePosition: google.maps.StrokePosition.CENTER});
        break;
        case 'inside':
          this.rectangle.setOptions({strokePosition: google.maps.StrokePosition.INSIDE});
        break;
        case 'outside':
          this.rectangle.setOptions({strokePosition: google.maps.StrokePosition.OUTSIDE});
        break;
      }
    }
  }

  _strokeWeightChanged(newValue, oldValue) {
    if (this.rectangle)
      this.rectangle.setOptions({strokeWeight: newValue});
  }

  _visibleChanged(newValue, oldValue) {
    if (this.rectangle)
      this.rectangle.setOptions({visible: newValue});
  }

  _zIndexChanged(newValue, oldValue) {
    if (this.rectangle)
      this.rectangle.setOptions({zIndex: newValue});
  }
}

window.customElements.define(GoogleMapRectangle.is, GoogleMapRectangle);

File: components/gmaps-flow/trunk/gmaps-flow-root/gmaps-flow/src/main/java/com/neotropic/flow/component/googlemap/GoogleMapRectangle.java
/*
 *  Copyright 2010-2020 Neotropic SAS <contact@neotropic.co>.
 *  
 *  Licensed under the Apache License, Version 2.0 (the "License"); 
 *  you may not use this file except in compliance with the License. 
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software 
 *  distributed under the License is distributed on an "AS IS" BASIS, 
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
 *  See the License for the specific language governing permissions and 
 *  limitations under the License.
 */
package com.neotropic.flow.component.googlemap;

import com.neotropic.flow.component.googlemap.GoogleMapEvent.RectangleBoundsChangedEvent;
import com.neotropic.flow.component.googlemap.GoogleMapEvent.RectangleClickEvent;
import com.neotropic.flow.component.googlemap.GoogleMapEvent.RectangleDblClickEvent;
import com.neotropic.flow.component.googlemap.GoogleMapEvent.RectangleDragEndEvent;
import com.neotropic.flow.component.googlemap.GoogleMapEvent.RectangleDragEvent;
import com.neotropic.flow.component.googlemap.GoogleMapEvent.RectangleDragStartEvent;
import com.neotropic.flow.component.googlemap.GoogleMapEvent.RectangleMouseDownEvent;
import com.neotropic.flow.component.googlemap.GoogleMapEvent.RectangleMouseMoveEvent;
import com.neotropic.flow.component.googlemap.GoogleMapEvent.RectangleMouseOutEvent;
import com.neotropic.flow.component.googlemap.GoogleMapEvent.RectangleMouseOverEvent;
import com.neotropic.flow.component.googlemap.GoogleMapEvent.RectangleMouseUpEvent;
import com.neotropic.flow.component.googlemap.GoogleMapEvent.RectangleRightClickEvent;
import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.ComponentEventListener;
import com.vaadin.flow.component.Synchronize;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.shared.Registration;
import elemental.json.Json;
import elemental.json.JsonObject;

/**
 *
 * @author Johny Andres Ortega Ruiz {@literal <johny.ortega@kuwaiba.org>}
 */
@Tag("google-map-rectangle")
@JsModule("./gmaps-rectangle.js")
public class GoogleMapRectangle extends Component {
    
    public GoogleMapRectangle() {
    }
    
    public GoogleMapRectangle(LatLngBounds bounds) {
        if (bounds != null)
            getElement().setPropertyJson(Constants.Property.BOUNDS, bounds.toJson());
    }
    //<editor-fold desc="Properties" defaultstate="collapsed">
    @Synchronize(property="bounds", value="rectangle-bounds-changed")
    public LatLngBounds getBounds() {
        return new LatLngBounds(
            (JsonObject) getElement().getPropertyRaw(Constants.Property.BOUNDS)
        );
    }
    
    public void setBounds(LatLngBounds bounds) {
        if (bounds != null)
            getElement().setPropertyJson(Constants.Property.BOUNDS, bounds.toJson());
        else
            getElement().setPropertyJson(Constants.Property.BOUNDS, Json.createNull());
    }
    
    public boolean getClickable() {
        return getElement().getProperty(Constants.Property.CLICKABLE, true);
    }
    
    public void setClickable(boolean clickable) {
        getElement().setProperty(Constants.Property.CLICKABLE, clickable);
    }
    
    public boolean getDraggable() {
        return getElement().getProperty(Constants.Property.DRAGGABLE, false);
    }
    
    public void setDraggable(boolean draggable) {
        getElement().setProperty(Constants.Property.DRAGGABLE, draggable);
    }
    
    public boolean getEditable() {
        return getElement().getProperty(Constants.Property.EDITABLE, false);
    }
    
    public void setEditable(boolean editable) {
        getElement().setProperty(Constants.Property.EDITABLE, editable);
    }
    
    public String getFillColor() {
        return getElement().getProperty(Constants.Property.FILL_COLOR);
    }
    
    public void setFillColor(String fillColor) {
        getElement().setProperty(Constants.Property.FILL_COLOR, fillColor);
    }
    
    public double getFillOpacity() {
        return getElement().getProperty(Constants.Property.FILL_OPACITY, Constants.Default.FILL_OPACITY);
    }
    
    public void setFillOpacity(double fillOpacity) {
        getElement().setProperty(Constants.Property.FILL_OPACITY, fillOpacity);
    }
    
    public String getStrokeColor() {
        return getElement().getProperty(Constants.Property.STROKE_COLOR);
    }
    
    public void setStrokeColor(String strokeColor) {
        getElement().setProperty(Constants.Property.STROKE_COLOR, strokeColor);
    }
    
    public double getStrokeOpacity() {
        return getElement().getProperty(Constants.Property.STROKE_OPACITY, Constants.Default.STROKE_OPACITY);
    }
    
    public void setStrokeOpacity(double strokeOpacity) {
        getElement().setProperty(Constants.Property.STROKE_OPACITY, strokeOpacity);
    }
    
    public String getStrokePosition() {
        return getElement().getProperty(Constants.Property.STROKE_POSITION);
    }
    
    public void setStrokePosition(StrokePosition strokePosition) {
        if (strokePosition != null)
            getElement().setProperty(Constants.Property.STROKE_POSITION, strokePosition.asString());
        else
            getElement().setPropertyJson(Constants.Property.STROKE_POSITION, Json.createNull());
    }
    
    public double getStrokeWeight() {
        return getElement().getProperty(Constants.Property.STROKE_WEIGHT, Constants.Default.STROKE_WEIGHT);
    }
    
    public void setStrokeWeight(double strokeWeight) {
        getElement().setProperty(Constants.Property.STROKE_WEIGHT, strokeWeight);
    }
    
    public boolean getVisible() {
        return getElement().getProperty(Constants.Property.VISIBLE, true);
    }
    
    public void setRectangleVisible(boolean visible) {
        getElement().setProperty(Constants.Property.VISIBLE, visible);
    }
    
    public double getZIndex() {
        return getElement().getProperty(Constants.Property.Z_INDEX, Constants.Default.Z_INDEX);
    }
    
    public void setZIndex(double zIndex) {
        getElement().setProperty(Constants.Property.Z_INDEX, zIndex);
    }
    //</editor-fold>
    //<editor-fold desc="Listeners" defaultstate="collapsed">
    public Registration addRectangleBoundsChangedListener(ComponentEventListener<RectangleBoundsChangedEvent> listener) {
        return addListener(RectangleBoundsChangedEvent.class, listener);
    }
    
    public Registration addRectangleClickListener(ComponentEventListener<RectangleClickEvent> listener) {
        return addListener(RectangleClickEvent.class, listener);
    }
    
    public Registration addRectangleDblClickListener(ComponentEventListener<RectangleDblClickEvent> listener) {
        return addListener(RectangleDblClickEvent.class, listener);
    }
    
    public Registration addRectangleDragListener(ComponentEventListener<RectangleDragEvent> listener) {
        return addListener(RectangleDragEvent.class, listener);
    }
    
    public Registration addRectangleDragEndListener(ComponentEventListener<RectangleDragEndEvent> listener) {
        return addListener(RectangleDragEndEvent.class, listener);
    }
    
    public Registration addRectangleDragStartListener(ComponentEventListener<RectangleDragStartEvent> listener) {
        return addListener(RectangleDragStartEvent.class, listener);
    }
    
    public Registration addRectangleMouseDownListener(ComponentEventListener<RectangleMouseDownEvent> listener) {
        return addListener(RectangleMouseDownEvent.class, listener);
    }
    
    public Registration addRectangleMouseMoveListener(ComponentEventListener<RectangleMouseMoveEvent> listener) {
        return addListener(RectangleMouseMoveEvent.class, listener);
    }
    
    public Registration addRectangleMouseOutListener(ComponentEventListener<RectangleMouseOutEvent> listener) {
        return addListener(RectangleMouseOutEvent.class, listener);
    }
    
    public Registration addRectangleMouseOverListener(ComponentEventListener<RectangleMouseOverEvent> listener) {
        return addListener(RectangleMouseOverEvent.class, listener);
    }
    
    public Registration addRectangleMouseUpListener(ComponentEventListener<RectangleMouseUpEvent> listener) {
        return addListener(RectangleMouseUpEvent.class, listener);
    }
    
    public Registration addRectangleRightClickListener(ComponentEventListener<RectangleRightClickEvent> listener) {
        return addListener(RectangleRightClickEvent.class, listener);
    }
    //</editor-fold>
}
